File: src/056MergeIntervals/056MergeIntervals.go
package main

import (
	"fmt"
	"sort"
)

/**
 * Definition for an interval.
 */
type Interval struct {
	Start int
	End   int
}

type ByStart []Interval

func (a ByStart) Len() int           { return len(a) }
func (a ByStart) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByStart) Less(i, j int) bool { return a[i].Start < a[j].Start }

func merge(intervals []Interval) []Interval {
	if len(intervals) == 0 {
		return nil
	}
	sort.Sort(ByStart(intervals))

	results := make([]Interval, 1)
	results[0] = intervals[0]
	for i := 1; i < len(intervals); i++ {
		last := len(results) - 1
		if intervals[i].Start >= results[last].Start &&
			intervals[i].Start <= results[last].End {
			if results[last].End < intervals[i].End {
				results[last].End = intervals[i].End
			}
		} else {
			results = append(results, intervals[i])
		}
	}
	return results
}

func main() {
	intervals := []Interval{
		{2, 6},
		{1, 3},
		{3, 4},
		{15, 18},
		{8, 10},
	}
	fmt.Printf("%v\n", merge(intervals))
}

File: src/056MergeIntervals/056MergeIntervals.rs
impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));
        
        let mut merged_intervals:Vec<Vec<i32>> = vec![];
        for interval in intervals {
            let len = merged_intervals.len();
            if len == 0 || interval[0] > merged_intervals[len-1][1] {
                merged_intervals.push(interval);
            }else{
                merged_intervals[len-1][1] = std::cmp::max(merged_intervals[len-1][1], interval[1]);
            }
        }
        
        merged_intervals
    }
}
